mantidproject/mantid
# Updated Error Reporter to handle long stack traces and messages better

**Description of work.**

This PR adds handling for long strings in the error reporter.

1. The additional user context box now has a label to display the max length
2. The buttons to send a report are disabled if the additional context text is too long
3. Long stack traces (>10,000), which were causing errors, are now shortened before sending the report to the server

**To test:**

It can be tricky to test the error reporter. The best method I found was to temporarily remove the `--single-process` argument from my PyCharm configuration and then run the Segfault algorithm. Alternatively, you could try following the instructions in test number 3 [here](https://developer.mantidproject.org/Testing/ErrorReporter-ProjectRecovery/ErrorReporterTesting.html).

Assuming you've got the error reporter to open,

- [x] Check that the new character limit label responds as expected to typing and copying in text
- Check that if you exceed the character limit...
  - [x] The label turns red
  - [x] The buttons to send the report are disabled
- Check that when the text is brought back within the character limit...
  - [x] The label turns black
  - [x] The buttons to send the report are enabled

It is trickier to test the case where an error has a stack trace >10,000 characters (let me know if you think of one without altering code). I have been altering the line 30 in `errorreports/presenter.py` to be `self._traceback = 'x' * (MAX_STACK_TRACE_LENGTH + 10)` and then running Segfault. Alternatively, you could change MAX_STACK_TRACE_LENGTH to something much smaller and evoke an error with test number 3 [here](https://developer.mantidproject.org/Testing/ErrorReporter-ProjectRecovery/ErrorReporterTesting.html).

Assuming you've got the error reporter open again,

- [x] Check that the full stack trace (without the `\n...\n`) is visible in the "Show more details" window
- [x] Write a message similar to "This is a test" and send the report
- [x] Check that in the console a warning that the stack trace has been reduced has been logged
- If you have access, go to the error_reports channel in slack and...
  - [x] Check the report is present
  - [x] Check that the stack trace is the reduced version (with the `\n...\n`)

Fixes #30402





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
